### PR TITLE
Fix vanilla giant jungle trees causing cascading generation

### PR DESCRIPTION
**Pull request by josephcsible**

### CODE REVIEW

**Comment by josephcsible** [comment]
Stack trace of the problem that this solves:
```
net.minecraft.world.chunk.Chunk.logCascadingWorldGeneration(Chunk.java:1541)
net.minecraft.world.chunk.Chunk.populate(Chunk.java:1003)
net.minecraft.world.chunk.Chunk.populate(Chunk.java:996)
net.minecraft.world.gen.ChunkProviderServer.provideChunk(ChunkProviderServer.java:157)
net.minecraft.world.World.getChunkFromChunkCoords(World.java:309)
net.minecraft.world.World.getChunkFromBlockCoords(World.java:304)
net.minecraft.world.World.getBlockState(World.java:910)
net.minecraft.world.World.observedNeighborChanged(World.java:580)
net.minecraft.world.World.updateObservingBlocksAt(World.java:483)
net.minecraft.world.World.markAndNotifyBlock(World.java:389)
net.minecraft.world.World.setBlockState(World.java:360)
net.minecraft.world.gen.feature.WorldGenerator.setBlockAndNotifyAdequately(WorldGenerator.java:36)
net.minecraft.world.gen.feature.WorldGenHugeTrees.growLeavesLayer(WorldGenHugeTrees.java:143)
net.minecraft.world.gen.feature.WorldGenMegaJungle.generate(WorldGenMegaJungle.java:50)
net.minecraft.world.biome.BiomeDecorator.genDecorations(BiomeDecorator.java:138)
net.minecraft.world.biome.BiomeDecorator.decorate(BiomeDecorator.java:88)
net.minecraft.world.biome.Biome.decorate(Biome.java:225)
net.minecraft.world.biome.BiomeJungle.decorate(BiomeJungle.java:81)
net.minecraft.world.gen.ChunkGeneratorOverworld.populate(ChunkGeneratorOverworld.java:464)
net.minecraft.world.chunk.Chunk.populate(Chunk.java:1015)
net.minecraft.world.chunk.Chunk.populate(Chunk.java:996)
net.minecraft.world.gen.ChunkProviderServer.provideChunk(ChunkProviderServer.java:157)
net.minecraft.server.MinecraftServer.initialWorldChunkLoad(MinecraftServer.java:344)
net.minecraft.server.integrated.IntegratedServer.loadAllWorlds(IntegratedServer.java:142)
net.minecraft.server.integrated.IntegratedServer.init(IntegratedServer.java:156)
net.minecraft.server.MinecraftServer.run(MinecraftServer.java:484)
java.lang.Thread.run(Thread.java:748)
```

**Comment by Gegy** [comment]
How could this effect mods using `WorldGenerator#setBlockAndNotifyAdequately`? Would it create unintended behavior in some cases?

**Comment by josephcsible** [comment]
I'd imagine not at all. I'm just changing "don't cause a block update, but do make adjacent observers tick" to "don't cause a block update or make adjacent observers tick", and in general you don't have observers in chunks that are still generating.

**Comment by bs2609** [comment]
Filed [MC-123860](https://bugs.mojang.com/browse/MC-123860) for this general issue (not setting flag 16 for world generation).

**Comment by williewillus** [comment]
How does the no observer flag fix this, does it load neighbor chunks to find observers?

**Comment by bs2609** [comment]
Yes, `getBlockState` is called for all adjacent positions, which causes chunks to be loaded if it happens on a chunk boundary.

File: patches/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
+--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java
++++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java
+@@ -33,7 +33,8 @@
+         }
+         else
+         {
+-            p_175903_1_.func_180501_a(p_175903_2_, p_175903_3_, 2);
++            int flag = net.minecraftforge.common.ForgeModContainer.fixVanillaCascading ? 2| 16 : 2; //Forge: With bit 5 unset, it will notify neighbors and load adjacent chunks.
```
Comment (bs2609) [inline patches/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java.patch:8]: If we're adding an extra line anyway, might as well put the comment on top for readability.

Comment (josephcsible) [reply patches/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java.patch:8]: I wanted to match how Lex did the exact same thing to [WorldGenLakes.java.patch](https://github.com/josephcsible/MinecraftForge/commit/0ad4218bc2d3a8cc13357d631fe1b2818e7a7e21#diff-65418d5bcae199ee6e3eba086ce3cc24).

Comment (LexManos) [reply patches/minecraft/net/minecraft/world/gen/feature/WorldGenerator.java.patch:8]: You could NOT add the second line. But it doesn't really matter. Its still clean enough.

**Pull request merged**